api/metrics/label_gatherer.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package metrics
+
+import (
+	"cmp"
+	"errors"
+	"fmt"
+	"slices"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+var (
+	_ MultiGatherer = (*prefixGatherer)(nil)
+
+	errDuplicateGatherer = errors.New("attempt to register duplicate gatherer")
+)
+
+// NewLabelGatherer returns a new MultiGatherer that merges metrics by adding a
+// new label.
+func NewLabelGatherer(labelName string) MultiGatherer {
+	return &labelGatherer{
+		labelName: labelName,
+	}
+}
+
+type labelGatherer struct {
+	multiGatherer
+
+	labelName string
+}
+
+func (g *labelGatherer) Register(labelValue string, gatherer prometheus.Gatherer) error {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	if slices.Contains(g.names, labelValue) {
+		return fmt.Errorf("%w: for %q with label %q",
+			errDuplicateGatherer,
+			g.labelName,
+			labelValue,
+		)
+	}
+
+	g.names = append(g.names, labelValue)
+	g.gatherers = append(g.gatherers, &labeledGatherer{
+		labelName:  g.labelName,
+		labelValue: labelValue,
+		gatherer:   gatherer,
+	})
+	return nil
+}
+
+type labeledGatherer struct {
+	labelName  string
+	labelValue string
+	gatherer   prometheus.Gatherer
+}
+
+func (g *labeledGatherer) Gather() ([]*dto.MetricFamily, error) {
+	gatheredMetricFamilies, err := g.gatherer.Gather()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, gatheredMetricFamily := range gatheredMetricFamilies {
+		if gatheredMetricFamily == nil {
+			continue
+		}
+
+		metrics := gatheredMetricFamily.Metric[:0]
```

Comment:
This removes `nil` metrics because I felt like it was weird to keep them around in a metrics family without adding the label...

Reply:
I looked into the prometheus library. It seems that they do not handle `nil` metrics families or `nil` metrics. So I'll remove our handling of them.